ipq807x: add Compex WPQ873

AnttiNykanen opened the pull request:
The Compex WPQ873 is a development board with two M.2 B-key slots for cellular modems.

### Device info
 - IPQ8072A SoC
 - 512MiB RAM
 - 256MiB NAND flash
 - 8MiB SPI NOR
 - 3x 1GigE ports
 - 1x 2.5GigE port
 - 2.4GHz/5GHz AX WLAN
 - 1x USB 3.0 port
 - 1x M.2 B-key socket with PCIe 3.0
 - 1x M.2 B-key socket with PCIe 2.0 and USB 3.0
 - 4x SIM card slots

### Installation

#### Prerequisites 
1. TFTP server
2. 3.3V USB to TTL cable for UART console
   2.54mm pitch 4-pin header for UART is readily provided on board, no modifications are necessary to access it
   TTL connector pinout: 2=TX, 3=RX, 4=GND
   Arrow marks pin 1 which is 3.3V
   Serial port settings: 115200 8N1 no flow control

The device will most likely ship with a QSDK-based firmware.

1. Power on device and interrupt u-boot to obtain u-boot CLI
2. set serverip to IP address of the TFTP server, for example:
        `setenv serverip 192.168.1.10`
3. Download image from TFTP server:
        `tftpboot 0x44000000 openwrt-ipq807x-generic-compex_wpq873-squashfs-factory.ubi`
4. Flash ubi image to both partitions and reset:
```
sf probe
imxtract 0x44000000 ubi
nand device 0
nand erase 0x0 0x3400000
nand erase 0x3c00000 0x3400000
nand write $fileaddr 0x0 $filesize
nand write $fileaddr 0x3c00000 $filesize
reset
```

Afterwards, you can use sysupgrade to flash new OpenWRT images.

AnttiNykanen commented:
I've been advised by company legal to not include the board WLAN file in this commit, since the licensing situation of binary blob files is problematic.

But I'm allowed to disclose that the correct board wlan file is:

```
$ sha256sum bdwlan.b292
a2951a29171c9f9c8d141edfaea5815a3336841b15b1f751a9e121caa19a3b7a  bdwlan.b292
$ du -sb bdwlan.b292
131072  bdwlan.b292
```

When this bdwlan file is used with the correct board-2 JSON (`bus=ahb,qmi-chip-id=0,qmi-board-id=255,variant=Compex-WPQ873`), to make a board-2.bin, WiFi works correctly on this board.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-wpq873.dts:
```diff
@@ -0,0 +1,488 @@
+&wifi {
+	status = "okay";
+
+	qcom,board_id = <0xFF>;
```
Remove this, its not used at all

AnttiNykanen replied:
Removed the qcom,board_id line

robimarko commented:
So far we have not been having any PR-s without the BDF-s as that leaves WLAN not working which I would argue is a major point of this SoC.

AnttiNykanen commented:
> So far we have not been having any PR-s without the BDF-s as that leaves WLAN not working which I would argue is a major point of this SoC.

Personally, I use this board to evaluate m.2 key-B cellular modem modules and don't use the WiFi at all. There are not that many boards that provide as versatile connectivity for 5G modem evaluation as this one does. 

As I mentioned, since I've done this at work, I must submit through the official process, and I've been advised against including the BDF binary blob in this PR. But I've also been told that there is no limit to how much I can advise other people on getting the bdwlan file committed. I'm committed on getting the WiFi working, as soon as I can figure out a plausible way to do it "cleanly".

AnttiNykanen commented:
> How much is the UART readily accessible? What is needed to access the UART? You may add in the commit description the fact you use a 1.8 V UART adapter or alternatively that you're using a 3.3V one since the target is tolerant to 3.3V.

The UART is 3.3V and there is a header readily provided on board. No modifications are necessary to access it. Have updated the commit message and PR description to reflect these facts.

robimarko commented:
Doesn't this device also have Bluetooth via QCA4024?

AnttiNykanen commented:
> Doesn't this device also have Bluetooth via QCA4024?

It would appear so - however I cannot get it to work.

The only UART enabled by the device tree in the stock fw is `serial@78B3000`, i.e. `blsp1_uart5`. Other UARTs are left disabled.

Even if I enable the UARTs @ 78af000 (_uart1) and 78B1000 (_uart3) in the dts and add kmod-bluetooth and bluez-utils, running `hciattach /dev/ttyMSMx bcsp` on either one of those always results in a timeout. Hciconfig also never shows anything. 

BTW, @outbackdingo has some of these boards as well and can provide the missing BDF for the wifi.

outbackdingo commented:
Hi Ive already spoken to @Ansuel yesterday about the bdf's and moving this forward... so he has what he needs now!

robimarko commented:
@AnttiNykanen I mean, it can be connected via USB as well, I have seen that before

AnttiNykanen commented:
> @AnttiNykanen I mean, it can be connected via USB as well, I have seen that before. Compex usually has good HW guides to show what is connected and where

I'll try to have a look again tomorrow, but unfortunately the [Hardware Guide](https://github.com/openwrt/openwrt/pull/12479#issuecomment-1549330582) doesn't say much except that GPIO21 is "BT_RESET". In the original DT, gpio21 appears to be quartz reset for blsp3_spi2.

Ansuel commented:
@AnttiNykanen took care of asking the BDF to Compex and uploaded to the BDF repository. Can you add to the DTS the correct variant so it can be loaded and add the related entry to the ipq-wifi package?

AnttiNykanen commented:
@Ansuel done

Ansuel commented:
@AnttiNykanen thanks a lot! Can you split the ipq wifi part to a separate commit bumping it to latest version?

You can follow this https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=69812bf8ed97e31f978cd9d05f6a181f6ade86c0 as an example

Commit:
ipq807x: add support for Compex WPQ873

The Compex WPQ873 is a development board with two M.2 B-key
 slots for cellular modems.

Device info:
 - IPQ8072A SoC
 - 512MiB RAM
 - 256MiB NAND flash
 - 8MiB SPI NOR
 - 3x 1GigE ports
 - 1x 2.5GigE port
 - 2.4GHz/5GHz AX WLAN
 - 1x USB 3.0 port
 - 1x M.2 B-key socket with PCIe 3.0
 - 1x M.2 B-key socket with PCIe 2.0 and USB 3.0
 - 4x SIM card slots
 - Bluetooth LE 5.0 (QCA4024)

Prerequisites
1) TFTP server
2) 3.3V USB to TTL cable for UART console
   2.54mm pitch 4-pin header for UART is readily provided on board, no modifications are necessary to access it
   TTL connector pinout: 2=TX, 3=RX, 4=GND
   Arrow marks pin 1 which is 3.3V
   Serial port settings: 115200 8N1 no flow control

The device will most likely ship with a QSDK-based firmware.

1. Power on device and interrupt u-boot to obtain u-boot CLI

2. set serverip to IP address of the TFTP server, for example:

        `setenv serverip 192.168.1.10`

3. Download image from TFTP server:

        `tftpboot 0x44000000 openwrt-ipq807x-generic-compex_wpq873-squashfs-factory.ubi`

4. Flash ubi image to both partitions and reset:

        `sf probe
         imxtract 0x44000000 ubi
         nand device 0
         nand erase 0x0 0x3400000
         nand erase 0x3c00000 0x3400000
         nand write $fileaddr 0x0 $filesize
         nand write $fileaddr 0x3c00000 $filesize
         reset`

Afterwards, you can use sysupgrade to flash new OpenWRT images.

Signed-off-by: Antti Nykänen <antti.nykanen@nokia.com>

Commit:
ipq-wifi: bump to latest git HEAD

0f73d32 ipq8074: update RegDB in new submitted BDF
a4cd21f ipq8074: add Compex WPQ873 BDF
c888dd0 qca-wireless: ipq40xx: Add BDFs for Eero Cento
6388ba9 ipq8074: update regdb for Netgear SXK80 BDF
77775d2 ipq8074: add Netgear SXK80

Signed-off-by: Antti Nykänen <antti.nykanen@nokia.com>

AnttiNykanen commented:
> @AnttiNykanen thanks a lot! Can you split the ipq wifi part to a separate commit bumping it to latest version?
> 
> You can follow this https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=69812bf8ed97e31f978cd9d05f6a181f6ade86c0 as an example

@Ansuel done

Ansuel commented:
@AnttiNykanen can you rebase this on top of main so I can flag this as merged (github limitation and i have to use git merge -ff-only to flag this as merged)

AnttiNykanen commented:
> @AnttiNykanen can you rebase this on top of main so I can flag this as merged (github limitation and i have to use git merge -ff-only to flag this as merged)

@Ansuel done
